### PR TITLE
Make method abstract and force subclasses to implement

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs [hunk 1]
```diff
@@ -5,43 +5,39 @@
 using System.Threading;
 using System.Threading.Tasks;
 
-namespace Microsoft.CodeAnalysis
+namespace Microsoft.CodeAnalysis;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs:8]: view with whitespace off.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs [hunk 2]
```diff
@@ -5,43 +5,39 @@
...
+internal partial class SolutionCompilationState
 {
-    internal partial class SolutionCompilationState
+    /// <summary>
+    /// Represents a change that needs to be made to a <see cref="Compilation"/>, <see cref="GeneratorDriver"/>, or both in response to
+    /// some user edit.
+    /// </summary>
+    private abstract partial class CompilationAndGeneratorDriverTranslationAction
     {
+        public abstract Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs:18]: i far preferred subclasses having to be intentional about what they should do here.  i already did something similar for CanUpdateCompilationWithStaleGeneratedTreesIfGeneratorsGiveSameOutput  previously.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction.cs:18]: now, the only non-abstract method is TryMergeWithPrior.  And that one is fine to stay virtual as it's only used as a perf optimization, not a correctness one.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction_Actions.cs [hunk 3]
```diff
@@ -55,10 +55,14 @@ internal sealed class TouchAdditionalDocumentAction(AdditionalDocumentState oldS
                 private readonly AdditionalDocumentState _newState = newState;
 
                 // Changing an additional document doesn't change the compilation directly, so we can "apply" the
-                // translation (which is a no-op). Since we use a 'false' here to mean that it's not worth keeping
-                // the compilation with stale trees around, answering true is still important.
+                // translation (which is a no-op). Since we use a 'false' here to mean that it's not worth keeping the
+                // compilation with stale trees around, answering true is still important.
                 public override bool CanUpdateCompilationWithStaleGeneratedTreesIfGeneratorsGiveSameOutput => true;
 
+                // Touching an additional document doesn't change the compilation in any way.  So can return as is.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction_Actions.cs:62]:
```suggestion
```

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Review by jasonmalinowski — COMMENTED** [review]
Looks good too.

**Pull request merged**